### PR TITLE
Avoid using pydantic compat layer where easily possible

### PR DESCRIPTION
**Pull request by felixfontein**
All other cases need v1 for easy linting (with extra checks).

### CODE REVIEW

File: src/antsibull_docs/schemas/app_context.py [hunk 1]
```diff
@@ -8,11 +8,10 @@
 import pydantic as p
 from antsibull_core.schemas.context import AppContext as CoreAppContext
 from antsibull_core.schemas.validators import convert_bool
-
-from antsibull_docs._pydantic_compat import Field
+from pydantic import Field
 
 #: Valid choices for the docs parsing backend
-DOC_PARSING_BACKEND_CHOICES_F = Field("auto", regex="^(auto|ansible-core-2\\.13)$")
+DOC_PARSING_BACKEND_CHOICES_F = Field("auto", pattern="^(auto|ansible-core-2\\.13)$")
```
Comment (gotmax23) [inline src/antsibull_docs/schemas/app_context.py:14]: While the Rust regex create used by pydantic v2 to validate regexes is more efficient than Python's re that was used before, we should switch these to use `Literal` instead, either here in another PR.

**Review by gotmax23 — COMMENTED** [review]
This looks good after a cursory look, but we should consider getting rid of the unnecessary regex usage for performance/simplicity reasons.

File: src/sphinx_antsibull_ext/schemas/ansible_links.py [hunk 2]
```diff
@@ -18,13 +18,14 @@ class AnsibleLink(p.BaseModel):
...
 
-    @p.root_validator()
-    # pylint:disable=no-self-argument
-    def one_of_url_and_ref(cls, values: dict[str, t.Any]) -> dict[str, t.Any]:
-        has_url = values.get("url")
-        has_ref = values.get("ref")
-        if has_url == has_ref:
-            raise ValueError("Exactly one of 'url' and 'ref' must be specified.")
+    @p.model_validator(mode="before")
+    @classmethod
```
Comment (gotmax23) [inline src/sphinx_antsibull_ext/schemas/ansible_links.py:22]: I don't think classmethod is strictly necessary from pydantic's side. Do one of our linters/type checkers complain if it's not there?

File: src/antsibull_docs/schemas/app_context.py (hunk 1 above)
Comment (felixfontein) [reply src/antsibull_docs/schemas/app_context.py:14]: I agree. But let's do that in another PR.

File: src/sphinx_antsibull_ext/schemas/ansible_links.py (hunk 2 above)
Comment (felixfontein) [reply src/sphinx_antsibull_ext/schemas/ansible_links.py:22]: I don't know whether it's really necessary (I haven't tried), but the (only) example in the pydantic documentation also uses it: https://docs.pydantic.dev/latest/concepts/validators/#model-validators

Comment (felixfontein) [reply src/sphinx_antsibull_ext/schemas/ansible_links.py:22]: (Also it avoids problems with other linters who don't know that pydantic does this.)

Comment (gotmax23) [reply src/sphinx_antsibull_ext/schemas/ansible_links.py:22]: > @field_validators are "class methods", so the first argument value they receive is the UserModel class, not an instance of UserModel. We recommend you use the @classmethod decorator on them below the @field_validator decorator to get proper type checking.

I suppose the same thing applies here.

Marking as resolved.

**Review by gotmax23 — APPROVED** [review]

**Comment by felixfontein** [comment]
@gotmax23 thanks for reviewing this!

**Comment by gotmax23** [comment]
Haha, was about to push that button

**Comment by gotmax23** [comment]
Thanks!

File: src/antsibull_docs/schemas/app_context.py (hunk 1 above)
Comment (felixfontein) [reply src/antsibull_docs/schemas/app_context.py:14]: Hmm, I tried replacing `doc_parsing_backend: str = DOC_PARSING_BACKEND_CHOICES_F` with `doc_parsing_backend: str = t.Literal["auto", "ansible-core-2.13"]`, and fed it a config file with an invalid value. With `t.Literal` it did **not** complain about the invalid value.

Comment (felixfontein) [reply src/antsibull_docs/schemas/app_context.py:14]: Hmm, and I tried it again with the CollectionMetadata schema from antsibull-core, and *there* it works... I have to try again :)

Comment (felixfontein) [reply src/antsibull_docs/schemas/app_context.py:14]: Ok, I did a stupid mistake. It works fine...

**Pull request merged**